src/cli/index.ts

```diff
@@ -0,0 +1,248 @@
+
+  const project = await writeStep(
+    `${chalk.bold("Importing")} ${inputType} project from path ${chalk.white(fullInputPath)}.`,
+    async () => {
+      let file: Buffer;
+      try {
+        file = fs.readFileSync(fullInputPath);
+      } catch (err) {
+        if (err instanceof Object && "code" in err) {
+          switch (err.code) {
+            case "ENOENT":
+              throw new StepError("File not found.");
+            case "EISDIR":
+              throw new StepError("Input path is a directory but should be a file.");
+          }
+        }
+        throw err;
+      }
+      const project = await Project.fromSb3(file);
+      return project;
+    }
+  );
+
+  switch (outputType) {
+    case "leopard": {
+      const leopard = await writeStep(`${chalk.bold("Converting")} project to ${chalk.white("Leopard")}.`, () => {
+        const leopard = project.toLeopard();
+        return leopard;
+      });
+
+      const fullOutputPath = resolve(process.cwd(), output);
+
+      await writeStep(`${chalk.bold("Exporting")} project to directory ${chalk.white(fullOutputPath)}.`, () => {
+        // First, make sure the output path is an empty directory (create it if it doesn't exist)
+        let existingFiles: string[];
+        try {
+          existingFiles = fs.readdirSync(fullOutputPath);
+        } catch (err) {
+          if (err instanceof Object && "code" in err) {
+            switch (err.code) {
+              case "ENOENT":
+                // Directory does not exist, create it
+                fs.mkdirSync(output, { recursive: true });
+                existingFiles = [];
+                break;
+              case "ENOTDIR":
+                throw new StepError("Output path is a file, not a directory.");
+              default:
+                throw err;
+            }
+          } else {
+            throw err;
+          }
+        }
+
+        if (existingFiles.length > 0) {
+          throw new StepError("Output directory is not empty.");
+        }
+
+        // Write the files
+        for (const [filename, contents] of Object.entries(leopard)) {
+          // Create directories as needed
+          fs.mkdirSync(resolve(fullOutputPath, filename, ".."), { recursive: true });
+
+          // Write the file
+          fs.writeFileSync(resolve(fullOutputPath, filename), contents);
+        }
+
+        for (const target of [project.stage, ...project.sprites]) {
+          for (const costume of target.costumes) {
+            const filename = `${target.name}/costumes/${costume.name}.${costume.ext}`;
+            const asset = Buffer.from(costume.asset as ArrayBuffer);
+            fs.mkdirSync(resolve(fullOutputPath, filename, ".."), { recursive: true });
+            fs.writeFileSync(resolve(fullOutputPath, filename), asset);
+          }
+
+          for (const sound of target.sounds) {
+            const filename = `${target.name}/sounds/${sound.name}.${sound.ext}`;
```

Comment:
Does it necessarily even make sense to hard-code asset paths, by the way? You're explicitly *not* hard-coding code filenames. Would it be possible, for each target, to get the dirname of its corresponding code filename, then put its `sounds` and `costumes` directories under that directory? I don't know if it's trivial to identify which sprite maps to which code filename though.

I do think this is a good idea beyond just "it's nice to not hard-code things" though, since it would make sure any special character handling that the Leopard converter does is respected and reflected here. It would be awkward if costumes and sounds ended up in a different folder than the code because the CLI doesn't match the logic that `toLeopard` uses.

Reply:
Yes, but this feels like a change that should be made to the `.toLeopard()` method. The way assets are currently handled is pretty gross to be honest. It might be nice to round up all the places that this method is being used (such as here and in the Leopard website source code) and examine the usage to design a better API.